shared/src/sources/balancer/pool_fetching.rs

```diff
@@ -228,29 +138,61 @@ impl StablePool {
             );
         }
         StablePool {
-            pool_id: pool_data.common.pool_id,
-            pool_address: pool_data.common.pool_address,
-            swap_fee_percentage,
-            amplification_parameter,
+            common: CommonPoolState {
+                pool_id: pool_data.common.pool_id,
+                pool_address: pool_data.common.pool_address,
+                swap_fee_percentage,
+                paused,
+            },
             reserves,
-            paused,
+            amplification_parameter,
         }
     }
 }
 
+impl BalancerPoolEvaluating for StablePool {
+    fn properties(&self) -> CommonPoolState {
+        self.common.clone()
+    }
+}
+
+pub struct FetchedBalancerPools {
+    pub stable_pools: Vec<StablePool>,
+    pub weighted_pools: Vec<WeightedPool>,
+}
+
+impl FetchedBalancerPools {
+    pub fn relevant_tokens(&self) -> HashSet<H160> {
+        let mut tokens = HashSet::new();
+        tokens.extend(
+            self.stable_pools
+                .iter()
+                .flat_map(|pool| pool.reserves.keys().copied().collect::<HashSet<_>>()),
```

Comment:
No need to collect into a temporary HashSet first
```suggestion
                .flat_map(|pool| pool.reserves.keys().copied()),
```
also 5 lines below

Reply:
Thanks! Someday I will get the hang of when this is or is not needed.

Reply:
The reason I *think* it is not needed is that we are collecting into a big hash set already. So any duplicates that the intermediate hash set would filter out would also be filtered out by the outer hash set.